### ISSUE TITLE
Enhancement: Executing untrusted SQL statements

### ISSUE THREAD

**Issue by benrb** [groue/GRDB.swift#1015]
I am looking to execute user-entered SQL without checking for safety or correctness - I know it sounds awful, but it's not!

This could be a mixture of various sorts of statements (`INSERT`, `SELECT`, `UPDATE` etc) all mixed together into one string.
Some of these statements can return result rows, others can't.

GRDB has separate function paths for SQL which select data vs those which don't. I would like to add a function something like this:

```swift
func executeUntrustedSQL(sql: String) throws -> [StatementResult]

enum StatementResult {
case .success(rows: [Row])
case .emptySuccess
case .failed(error: DatabaseError)
}
```

This could execute a bunch of statements of varying types and return a collection of results. For example:

``` SQL
CREATE TABLE t (id INTEGER PRIMARY KEY);
INSERT INTO t DEFAULT VALUES;
SELECT * FROM t;
INSERT INTO notT DEFAULT VALUES;
SELECT * FROM t;
```

Could return:

```swift
[
.emptySuccess,
.emptySuccess,
.success(rows),
.failed(error),
.success(rows)
]
```

Would there be interest in incorporating this into GRDB if I were to write it?
If yes, how do you feel about the proposed API?

---
Further notes:
- I believe only a single querying function would be required here. 
  - It makes sense to only support `String` inputs because this is targeted at SQL you do not own. For SQL you provide, you should be able to use the more specific `execute()` or `fetchOne()/fetchAll()` functions.
  - The `.success` case should be returning an array of `Row` objects. Again, if you know the specific type of result ahead of time, one of the other query interfaces would probably be better suited.
- `executeUntrustedSQL()` would catch and return errors from invalid SQL, but might still throw errors from other contexts, so remains marks as `throws`
- There are other things that could be good to add here, such as including the SQL for each query in the returned `StatementResult`, but the above is a more minimal addition.

**Comment by groue** [comment]
Hello @benrb,

You're looking after a Swift version of [`sqlite3_exec`](https://www.sqlite.org/c3ref/exec.html). I'd be interested in a Swift version of this function.

Your suggested solution presents a few problems that we should aim at avoiding:

- It distinguishes `emptySuccess` from `.success([])` for no known reason.
- It keeps on executing statements after the first error, which is bad. Some users will want to execute statements which are _preconditions_ for subsequent statements, and those users will be extremely unhappy to execute SQL statements on a wrong state of the database.
- It builds row arrays, but row arrays are very memory hungry when there are many rows. Think of users who want to have use this api but will fetch thousands and millions of rows.
- It does not accept arguments, as `execute(sql:arguments:)` does.

All in all, if you want to keep on investigating this topic, I strongly suggest you look at `sqlite3_exec` and try to see if it would fit your needs (chances are that it does, if only you had a Swift version of it). Unless I miss something, you are likely to model a function which only returns a cursor of rows.

Last random nits: don't say "untrusted" SQL.  It's just SQL. Your feature is no more dangerous than `execute(sql:arguments:)`, it just happens to emits rows as well.

**Comment by groue** [comment]
We won't use `Result`. Nearly all GRDB methods are throwing instead. Consider `DatabaseError` is not the only kind of error that can be thrown from database accesses. In particular, library users can throw custom errors when they cancel a transaction through TransactionObserver. There may be other opportunities for non-DatabaseError failures that I do not remember at present time. But even if GRDB would only throw `DatabaseError`, I'd still prefer throwing methods over `Result`. The language ergonomics favors throwing methods so much that Result is just not on the table.

I don't see why `[RowCursor?]`. `sqlite3_exec` does not distinguish the source of rows, so I don't know why we should do it. It looks one single cursor is enough.

Now users will also need arrays and sets. And users will also want to fetch other types than `Row`.

So I was actually thinking of allowing multiple statements in all fetching methods that accept raw SQL:

```swift
let rows = try Row.fetchCursor(db, sql: "CREATE ...; INSERT ...; SELECT ...", arguments: ...)
```

We have a precedent with `execute(sql:arguments:)`.

And users who want to make sure they run a single statement can use prepared statements.

Would this fit your use case?

---

I've started to look at this feature. It is an opportunity to remove the SelectStatement/UpdateStatement dichotomy.

I meet a difficulty, though. Statements like `DELETE FROM ...` will become easier to run through cursors. Such statements must run with a specific [sqlite authorizer](https://www.sqlite.org/c3ref/set_authorizer.html) in order to prevent the [truncate optimization](https://www.sqlite.org/lang_delete.html#the_truncate_optimization) whenever there exists an active [transaction observer](https://github.com/groue/GRDB.swift/blob/master/README.md#transactionobserver-protocol) which tracks deletions. And the tricky part is that cursor invocations can be interleaved with other database actions, unlike `execute` which fully consumes a statement before allowing other database actions. Juggling with authorizers on each statement step would be bad for performance.

**Comment by benrb** [comment]
I believe that an array of cursors makes the proposed new function much more useful that a single cursor. When pitching for this I wasn't thinking in terms of a direct replacement for `sqlite3_exec()` but somewhere between that and what we already have.

When executing a series of statements, they might not all have the same-shaped results. My app doesn't always know what SQL is being provided for execution, but does need to be able to present the results.

For example, when executing the following:
```sql
SELECT id, name FROM people;
SELECT id, name FROM staff;
SELECT home FROM addresses;
```
I want to be able to show three separate sets of results. If we returned a single cursor, I would have to detect when the results changed in from one statement to the next and handle that which may not be easy. 

On the other hand if each statement returned its own cursor I could process them separately. Another GRDB user who doesn't mind having the results all lumped together can enumerate over the collection of cursors easily and get the result they want too.

The changes you suggest - while positive - would likely not fulfil my use case.

However, the thought of merging `SelectStatement` and `UpdateStatement` would help me create an extension to fulfil my own requirements without having to add it to the core framework. More on this below.

This has possibly expanded to more than I think I could reasonably contribute. I am obviously not as familiar with the GRDB codebase as you are and would be concerned about breaking things or working at cross purposes to your vision of the framework. I am happy to help wherever possible though.

---

I have a rough version of the initial pitch mocked up and working, though messy and not properly tested. It still uses `Result` types rather than the throwing errors. I can change this later. The function naming is particularly bad - this is just a demonstration of my thoughts!

You can see it in benrb/GRDB.swift@4664fd7e65439eda1c3e93c9d419ff8c0ebdc900: 

I refactored `Statement.prepare()` and added `Statement.prepareSomeTypeOfStatement` to prepare an `SQLiteStatement` and then return either a `SelectStatement` or `UpdateStatement` depending on its effects.

Then in Database+Statements.swift I added a new `executeMultiple()` function which prepares statements using the above and calls either `Row.fetchCursor()` or `updateStatement.execute()` depending on which type of statement was prepared.

I think that if `makeSomeStatement()` in the above commit was public API, then I could likely build my ideal function in an extension without asking to add it to the core. At the moment I don't have a way to prepare a statement without knowing the type in advance.

**Comment by groue** [comment]
> I believe that an array of cursors makes the proposed new function much more useful that a single cursor. When pitching for this I wasn't thinking in terms of a direct replacement for `sqlite3_exec()` but somewhere between that and what we already have.

OK let's discuss :-)

> When executing a series of statements, they might not all have the same-shaped results. My app doesn't always know what SQL is being provided for execution, but does need to be able to present the results.
> 
> For example, when executing the following:
> 
> ```sql
> SELECT id, name FROM people;
> SELECT id, name FROM staff;
> SELECT home FROM addresses;
> ```
> 
> I want to be able to show three separate sets of results. If we returned a single cursor, I would have to detect when the results changed in from one statement to the next and handle that which may not be easy.
> 
> On the other hand if each statement returned its own cursor I could process them separately. Another GRDB user who doesn't mind having the results all lumped together can enumerate over the collection of cursors easily and get the result they want too.
> 
> The changes you suggest - while positive - would likely not fulfil my use case.

Would you mind describing your use case (not just the GRDB api you want, but what you want to do in the end)?

> However, the thought of merging `SelectStatement` and `UpdateStatement` would help me create an extension to fulfil my own requirements without having to add it to the core framework. More on this below.
> 
> This has possibly expanded to more than I think I could reasonably contribute. I am obviously not as familiar with the GRDB codebase as you are and would be concerned about breaking things or working at cross purposes to your vision of the framework. I am happy to help wherever possible though.

No problem. You help a lot. Such discussion is very useful. My job here is partially to know the GRDB code base, as you say, but mainly to make sure any new API can fulfill as many use cases as possible, and remain consistent with the rest of the framework.

You see this line of reasoning when I wrote above that some users will want to fetch arrays, or sets, or strings. It is not because YOU want to fetch heterogeneous sets of results that other users must pay for the complexity of your app. In such a world, GRDB would not have shipped with `String.fetchXXX` or `Player.fetchXXX` methods on top of `Row.fetchXXX`. We must use our imagination, foresee what other users could need, and *expand* your initial scenario as much as possible while preserving a simple and consistent mental model.

> I have a rough version of the initial pitch mocked up and working, though messy and not properly tested. It still uses `Result` types rather than the throwing errors. I can change this later. The function naming is particularly bad - this is just a demonstration of my thoughts!
> 
> You can see it in [benrb/GRDB.swift@4664fd7](https://github.com/benrb/GRDB.swift/commit/4664fd7e65439eda1c3e93c9d419ff8c0ebdc900):

Yeah, that's cool! I'm glad you could make progress without waiting for the feature to stabilize 👍 

---

My morning shower thoughts led me to yet another possible solution:

1. Keep on merging SelectStatements and UpdateStatements (they just become type aliases to Statement)
2. Expose `Statement.isReadOnly` so that users can more or less expect a statement to return rows or not. I'll have a better look at the kind of meta data SQLite provides on prepared statements. Such flag would help "replacing" the previous distinct classes.
3. Expose a new api which returns a cursor of statements from an SQL string.

You could write something like:

```swift
// A cursor of statements
let statements = db.allStatements(sql: "...")
while let statement = try statements.next() {
    let rows = try Row.fetchAll(statement)
    // use rows
}

// A single cursor of all rows (more or less matches sqlite3_exec)
let rows = db.allStatements(sql: "...").map(Row.fetchCursor).joined()
```

`allStatemens` would return a cursor of statements, because we can't return an array or a sequence:

```sql
-- Can't return an array of statements because
-- the second one won't compile until the first is run.
CREATE TABLE player (...);
INSERT INTO player;

-- Can't return a sequence of statements because
-- we need to handle errors at any time (here in the
-- second "statement":
CREATE TABLE player (...);
NOPE I WON'T COMPILE;
```

(A sequence of `Result` would only partially model such a cursor, because a sequence of Result is able to output a success after a failure. Only cursors make sure a failure marks the end of the iteration.)

**Comment by benrb** [comment]
> Would you mind describing your use case (not just the GRDB api you want, but what you want to do in the end)?

Sure. I want to present a UI for running arbitrary statements on a database. I want to present a freeform text view, then below that a tabbed collection of results from any statements run from that box. [Something like this](https://dev.mysql.com/doc/visual-studio/en/images/m4vs-sql-editor-layout.png), but with multiple tabs.

> No problem. You help a lot. Such discussion is very useful. My job here is partially to know the GRDB code base, as you say, but mainly to make sure any new API can fulfill as many use cases as possible, and remain consistent with the rest of the framework.
>
> You see this line of reasoning when I wrote above that some users will want to fetch arrays, or sets, or strings. It is not because YOU want to fetch heterogeneous sets of results that other users must pay for the complexity of your app. It is also not because YOU do not need statement arguments that we must prevent anyone from providing some. We must use our imagination, foresee what other users could need, and _expand_ your initial scenario as much as possible while preserving a simple and consistent mental model. This prevents a lot of frustration for everybody.

👍 That is a thoughtful explanation, thank you. The solution you propose below is much more capable while being simpler to understand.

> My morning shower thoughts led me to yet another possible solution:
> 
> 1. Keep on merging SelectStatements and UpdateStatements (they just become type aliases to Statement)
> 2. Expose `Statement.isReadOnly` so that users can more or less expect a statement to return rows or not. I'll have a better look at the kind of meta data SQLite provides on prepared statements. Such flag would help "replacing" the previous distinct classes.
> 3. Expose a new api which returns a cursor of statements from an SQL string.
> 

This looks like a good answer. It certainly appears to meet the criteria you set out and would definitely work for me.

**Comment by groue** [comment]
Thank you very much, @benrb, I'm very glad we're converging 🙂 I know what to aim for now 🚀 

And I sure understand now why you want to split statements!

**Comment by groue** [comment]
OK the [dev/merge-statements](https://github.com/groue/GRDB.swift/tree/dev/merge-statements) branch contains both the merging of statement types and `allStatements()`. I know have to find how and where it should be documented.

**Comment by groue** [comment]
#1019 was merged into the development branch

**Comment by benrb** [comment]
Thank you @groue, that looks fantastic! I am sure this will be a useful change for both myself and others.

**Comment by groue** [comment]
The new api has shipped in [v5.9.0](https://github.com/groue/GRDB.swift/tree/v5.9.0).
